server.js

```diff
@@ -6287,6 +6287,92 @@ cache(function(data, match, sendBadge, request) {
   });
 }));
 
+// MicroBadger integration.
+camp.route(/^\/microbadger\/(image-size|layers)\/([^/]+)\/([^/]+)\/?([^/]*)\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  var type = match[1];
+  var user = match[2];
+  var repo = match[3];
+  var tag = match[4];
+  var format = match[5];
+  if (user === '_') {
+    user = 'library';
+  }
+  var path = user + '/' + repo;
+  var url = 'https://api.microbadger.com/v1/images/' + path;
+
+  var badgeData = getBadgeData(type, data);
+  if (type === 'image-size') {
+    badgeData.text[0] = getLabel('image size', data);
+  }
+
+  var options = {
+    method: 'GET',
+    uri: url,
+    headers: {
+      'Accept': 'application/json'
+    }
+  };
+  request(options, function(err, res, buffer) {
+    if (res && res.statusCode === 404) {
+      badgeData.text[1] = 'not found';
+      sendBadge(format, badgeData);
+      return;
+    }
+
+    if (err != null || !res || res.statusCode !== 200) {
+      badgeData.text[1] = 'inaccessible';
+      sendBadge(format, badgeData);
+      return;
+    }
+
+    try {
+      var parsedData = JSON.parse(buffer);
+      var image;
+
+      if (tag) {
+        if (!parsedData['Versions']) {
+          badgeData.text[1] = 'not found';
+          sendBadge(format, badgeData);
+          return;
+        }
+
+        for (var i = 0; !image && i < parsedData['Versions'].length; i++) {
+          var versionJson = parsedData['Versions'][i];
+          for (var j = 0; !image && j < versionJson['Tags'].length; j++) {
+            var tagJson = versionJson['Tags'][j];
+            if (tag === tagJson.tag) {
+              image = versionJson;
+            }
+          }
+        }
+      } else {
+        image = parsedData;
+      }
+
```

Comment:
There's a chunk of logic in the middle here which parses the api response to find the latest tag. Could this be extracted to a helper?

To follow convention, you would create `/lib/microbadger-helpers.js` and move this bit of logic to a function `getLatestTag()` or `getLatestVersionTag()` or something. That would reduce the amount of branching/logic in this function making it easier to read and by factoring it out it would be easier to test. You could add few unit tests for it in `/lib/microbadger-spec.js`.

If you have a look through `/lib` there are some examples following this pattern that you can crib from.

Reply:
Sure, I'll take a look! Thank you for the tip. I've just seen several other services with similar complexity code but I would like to break off that code too.